File: 556 next greater element III.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jul 11 18:38:08 2018

@author: tzlmyq
"""



class Solution(object):
    def nextGreaterElement(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n <= 9:
            return -1
        ls = list(str(n))
        if str(n) == "".join(sorted(ls, reverse = True)):
            return -1
        else:
            for i in range(len(ls)-1, -1, -1):
                if ls[i] > ls[i-1]:
                    
                    res = "".join(ls[:i-1])
                    temp = ls[i-1]
                    left = ls[i:]
                    left.sort()
                    for j in range(len(left)):
                        if left[j] > temp:
                            temp, left[j] = left[j], temp
                            break
                    res += temp
                    res += "".join(left)
                    return int(res) if int(res) < 2**31-1 else -1
                
                
solution = Solution()
n = 12
print solution.nextGreaterElement(n)

File: 556. Next Greater Element III.cpp
class Solution {
public:
    int nextGreaterElement(int n) {
        vector<int> v;
        long temp = long(n);
        while(n)
        {
            v.push_back(n % 10);
            n = n / 10;
        }
        int left = 0, right = 0;
        while(right < v.size() - 1)
        {
            if(v[right] > v[right + 1])
                break;
            else
                right++;
        }
        if(right == v.size() - 1)
            return -1;
        right++;
        int index = right - 1;
        while(left < right)
        {
            if(v[left] > v[right] && v[left] < v[index])
                index = left;
            left++;
        }
        swap(v[index], v[right]);
        sort(v.begin(), v.begin() + right, std::greater<>());
        temp = 0;
        while(!v.empty())
        {
            temp = temp * 10 + v.back();
            v.pop_back();
        }
        if(temp > INT_MAX)
            return -1;
        return temp;
    }
};
